include/mega/command.h

```diff
@@ -84,8 +84,83 @@ class MEGA_API Command
     void closeobject();
     int elements();
 
-    virtual void procresult();
-
+    enum Outcome {  CmdError,            // The reply was an error, already extracted from the JSON.  The error code may have been 0 (API_OK)
+                    //CmdActionpacket,     // The reply was a cmdseq string, and we have processed the corresponding actionpackets
+                    CmdArray,            // The reply was an array, and we have already entered it
+                    CmdObject,           // the reply was an object, and we have already entered it
+                    CmdItem };           // The reply was none of the above - so a string
+
+    struct Result
+    {
+        Outcome mOutcome = CmdError;
+        Error mError = API_OK;
+        Result(Outcome o, Error e = API_OK) : mOutcome(o), mError(e) {}
+
+        bool succeeded()
+        {
+            return mOutcome != CmdError || error(mError) != API_OK;
+        }
+
+        Error errorGeneral()
+        {
+            // for the general case- command could result in an error, actionpacket, or JSON to process
+            return mOutcome == CmdError ? mError : Error(API_OK);
+        }
+
+        bool hasJsonArray()
+        {
+            // true if there is JSON Array to process (and we have already entered it) (note some commands that respond with cmdseq plus JSON, so this can happen for actionpacket results)
+            return mOutcome == CmdArray;
+        }
+
+        bool hasJsonObject()
+        {
+            // true if there is JSON Object to process (and we have already entered it) (note some commands that respond with cmdseq plus JSON, so this can happen for actionpacket results)
+            return mOutcome == CmdObject;
+        }
+
+        bool hasJsonItem()
+        {
+            // true if there is JSON to process but it's not an object or array (note some commands that respond with cmdseq plus JSON, so this can happen for actionpacket results)
+            return mOutcome == CmdItem;
+        }
+
+        // convenience function for commands that should only return numeric or actionpacket, consider anything else EINTERNAL
+        //Error errorResultOrActionpacket()
+        //{
+        //    return mOutcome == CmdError ? mError : Error(mOutcome == CmdActionpacket ? API_OK : API_EINTERNAL);
+        //}
+
+        Error errorOrOK()
+        {
+            assert(mOutcome == CmdError);
+            return mOutcome == CmdError ? mError : Error(API_EINTERNAL);
+        }
+
+        //bool wasErrorOrActionpacket()
+        //{
+        //    return mOutcome == CmdError || mOutcome == CmdActionpacket;
+        //}
+
+        bool wasErrorOrOK()
+        {
+            return mOutcome == CmdError;
+        }
+
+        bool wasError(error e)
+        {
+            return mOutcome == CmdError && error(mError) == e;
```

Comment:
Is there any specific reason for the casting? I think the compiler should decide to use the `Error::operator::error()` automatically, in cases like this one.